### PR TITLE
Add option to run LLVM optimisation passes

### PR DESCRIPTION
**Pull request by pramodk**
- update CLI argument from --llvm to llvm --ir --opt
  - --ir runs CodegenLLVMVicitor and emits LLVM IR
  - --opt runs basic LLVM optimisation passes
  - update simple test to check optimisation passes

@georgemitenkov : I was playing with LLVM passes to see if/how they work. I have opened this PR to your  #470 just for simplicity (to easily see the diff). This is just a draft PR anyway! If you think it's useful then we can use it after #470 is finalised or keep it for `later`. 

What this does is: current output **without** optimisations enabled:

```
$ ./bin/nmodl ../test/integration/mod/procedure.mod llvm --ir
[NMODL] [info] :: Processing ../test/integration/mod/procedure.mod
[NMODL] [info] :: Running symtab visitor
[NMODL] [info] :: Running CVode to cnexp visitor
[NMODL] [info] :: Running LOCAL to ASSIGNED visitor
[NMODL] [info] :: Running code compatibility checker
[NMODL] [info] :: Running verbatim rename visitor
[NMODL] [info] :: Running KINETIC block visitor
[NMODL] [info] :: Running STEADYSTATE visitor
[NMODL] [info] :: Parsing Units
[NMODL] [info] :: Running cnexp visitor
[NMODL] [info] :: Running C backend code generator
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 15.23 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 18.28 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 19.27 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 24.25 to arg_v
[NMODL] [info] :: Running LLVM backend code generator
; ModuleID = 'procedure'
source_filename = "procedure"

define void @hello_world() {
  ret void
}

define void @simple_sum(double %x1, double %y2) {
  %x = alloca double
  store double %x1, double* %x
  %y = alloca double
  store double %y2, double* %y
  %z = alloca double
  %1 = load double, double* %y
  %2 = load double, double* %x
  %3 = fadd double %2, %1
  store double %3, double* %z
  ret void
}

define void @complex_sum(double %arg_v1) {
  %arg_v = alloca double
  store double %arg_v1, double* %arg_v
  %alpha = alloca double
  %beta = alloca double
  %sum = alloca double
  %1 = load double, double* %arg_v
  %2 = fadd double %1, 4.000000e+01
  %3 = fsub double -0.000000e+00, %2
  %4 = fmul double 1.000000e-01, %3
  store double %4, double* %alpha
  %5 = load double, double* %arg_v
  %6 = fadd double %5, 6.500000e+01
  %7 = fsub double -0.000000e+00, %6
  %8 = fdiv double %7, 1.800000e+01
  %9 = fmul double 4.000000e+00, %8
  store double %9, double* %beta
  %10 = load double, double* %beta
  %11 = load double, double* %alpha
  %12 = fadd double %11, %10
  store double %12, double* %sum
  ret void
}

define void @loop_function(double %arg_v1) {
  %arg_v = alloca double
  store double %arg_v1, double* %arg_v
  %i = alloca double
  store double 0.000000e+00, double* %i
  ret void
}
```

And with optimisations enabled:

```
$ ./bin/nmodl ../test/integration/mod/procedure.mod llvm --ir --opt
[NMODL] [info] :: Processing ../test/integration/mod/procedure.mod
[NMODL] [info] :: Running symtab visitor
[NMODL] [info] :: Running CVode to cnexp visitor
[NMODL] [info] :: Running LOCAL to ASSIGNED visitor
[NMODL] [info] :: Running code compatibility checker
[NMODL] [info] :: Running verbatim rename visitor
[NMODL] [info] :: Running KINETIC block visitor
[NMODL] [info] :: Running STEADYSTATE visitor
[NMODL] [info] :: Parsing Units
[NMODL] [info] :: Running cnexp visitor
[NMODL] [info] :: Running C backend code generator
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 15.23 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 18.28 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 19.27 to arg_v
[NMODL] [warning] :: RenameVisitor :: Renaming variable v at 24.25 to arg_v
[NMODL] [info] :: Running LLVM backend code generator
[NMODL] [info] :: Running LLVM optimisation passes
; ModuleID = 'procedure'
source_filename = "procedure"

define void @hello_world() {
  ret void
}

define void @simple_sum(double %x1, double %y2) {
  ret void
}

define void @complex_sum(double %arg_v1) {
  ret void
}

define void @loop_function(double %arg_v1) {
  ret void
}
```

### CODE REVIEW

**Comment by georgemitenkov** [comment]
This looks good!  Great to see that everything seems to be working :) This will fit nicely when we start integration testing with LLVM's execution engine. I am planning to start working on this so that we can fully execute functions/procedures.

**Comment by pramodk** [comment]
Yup! It will be more fun once we have FUNCTION return types correctly working form your PR #470. Then we can see optimisations correctly happening! 

I will rebase this once #470 will be merged and then this should be good to merge.

**Comment by georgemitenkov** [comment]
Yes, that's good!

**Comment by pramodk** [comment]
@georgemitenkov : let me know if I should merge this (after your review/approval)

**Review by georgemitenkov — APPROVED** [review]
LGTM! One comment - maybe we should separate optimisation tests from IR codegen tests.

**Comment by pramodk** [comment]
> One comment - maybe we should separate optimisation tests from IR codegen tests.

Agree! I will fix this tomorrow and then merge.

**Pull request merged**
